### ISSUE TITLE
Improve Interactive States for Bokeh Widgets

### ISSUE THREAD

**Issue by smeragoel** [bokeh/bokeh#14410]
### Problem description

[Bokeh widgets](https://docs.bokeh.org/en/latest/docs/user_guide/interaction/widgets.html#widgets-and-dom-elements) currently have inconsistent or missing interactive states (hover, focus, active, disabled). This impacts usability and accessibility especially  for keyboard and mouse interactions.

### Feature description

1. Improve interaction states (hover, focus, active, disabled) across all widgets.
2. Address accessibility concerns - missing focus indicators, minimum contrast requirements not met, etc.
3. Establish a design system in Figma to document styles and refine the design where needed as an added benefit.

### Potential alternatives

1. Keeping the current interaction states as-is (not ideal).
2. Making improvements per widget instead of a systematic approach. However, this approach can lead to fragmented and inconsistent UX.

### Additional information

1. Scope & prioritization: [Widget interaction design prioritization - Spreadsheet](https://docs.google.com/spreadsheets/d/1yVkWEbsjxjH7Wt4iTWuZB6gjHY3-gpNJ5Fz-wtkS62Y/edit?usp=sharing)
2. Accessibility guidelines (WCAG) and best practices will inform improvements.

CC: @trallard

**Comment by smeragoel** [comment]
### Progress Update on Widget Design Work

I'm still actively working on the widget designs, but if anyone would like to follow along or leave feedback, here's the Figma file where the work is happening: [Bokeh Design System](https://www.figma.com/design/3xk3QObj3XaALcHHyz8nmE/Bokeh-Design-System?node-id=79-340&t=vIj8hYohXXIShcAn-1)

The file is organized into 4 main sections, visible via the left-hand panel in Figma (you might need to scroll or expand the **Pages** section depending on your viewport):

<details>

<summary>Screenshot of Pages Section</summary>

![Image](https://github.com/user-attachments/assets/0a249298-dad7-460d-93b3-f78f43d0de55)


</details>

1. **Ready for Dev**: Finalized designs that have been discussed and are ready to move into development.
2. **Needs Review**: Design work is complete but pending discussioon. Feel free to leave comments and suggestions here.
3. **Work in progress**: Actively being worked upon, so things might look wonky. Comments are still welcome but please remember that designs are not final.
4. Staging Area: Used for maintaining the file and broader design system structure. Mostly relevant for designers.

Always happy to hear your thoughts. Feel free to poke around and leave comments. Thanks! 😊

**Comment by trallard** [comment]
TODO: @pavithraes @trallard and @smeragoel  to realign on widgets scope and create a checklist for implementation work

**Comment by smeragoel** [comment]
All scoped widget components have now been designed and are available in the Figma file: [Bokeh Design System](https://www.figma.com/design/3xk3QObj3XaALcHHyz8nmE/Bokeh-Design-System?node-id=278-3156&t=Wf7nyUa0zZMeny8p-1)

I've moved everything into the **Ready for Dev** section. The designs have either been reviewed already or are in a good state to move forward, but if anything looks off or needs more discussion, feel free to flag it. Would love to get a final round of eyes on this before handoff!

Thanks again! 😊

Updating to add that I moved the [MultiChoice](https://www.figma.com/design/3xk3QObj3XaALcHHyz8nmE/Bokeh-Design-System?node-id=718-213) widget back to WIP because I remembered I needed to tweak it based on #14486

**Comment by smeragoel** [comment]
### Discussion Summary
During today's team sync, we discussed the [MultiChoice](https://www.figma.com/design/3xk3QObj3XaALcHHyz8nmE/Bokeh-Design-System?node-id=718-213) widget issue. Here's a summary:

1. For context, during [widget prioritisation](https://docs.google.com/spreadsheets/d/1yVkWEbsjxjH7Wt4iTWuZB6gjHY3-gpNJ5Fz-wtkS62Y/edit?gid=0#gid=0), we discussed that Multichoice and [MultiSelect](https://docs.bokeh.org/en/latest/docs/user_guide/interaction/widgets.html#multiselect) both have similar functions but MultiSelect does not have the best UX, so it was decided that we would work on just MultiChoice and recommend using that instead of MultiSelect.

2. The MultiChoice widget has some interactions that could be improved.
    1. No visual change on hover or focus (anywhere on the widget)
    2. When you click the pill or the 'text box', it triggers the dropdown select menu. When you click on the 'x', it removes that pill. Clicking on the pill shouldn't trigger the dropdown.
    3. This is a good end-goal for this widget: https://mui.com/material-ui/react-autocomplete/#multiple-values

3. In #14486, it's pointed out that it isn't clear how to interact with the MultiSelect widget, and that it takes up a lot of screen real estate when multiple options are selected, the latter also being a problem with the MultiChoice widget. I think the redesign will help in the first issue, but the new widget will still take up a lot of space. A close example of the suggested implementation would be this: https://mui.com/material-ui/react-select/#checkmarks.

So, we have three widgets which try to accomplish a similar goal, but differ in implementation, UX quality and interaction:
1. The native MultiSelect
2. The (slightly improved) MultiChoice
3. A new Checkmark-based widget

### Next Steps
1. We will leave the native MultiSelect as is. It can be safely assumed that it's not easy to customise this widget, especially to the degree we want.
2. We will improve the interaction states of the current MultiChoice widget to bring it closer to https://mui.com/material-ui/react-autocomplete/#multiple-values
3. The new Checkmark-based widget would be a good addition to Bokeh's widget collection, but its work will be tracked separately.

**Comment by smeragoel** [comment]
Since the last update on this issue, we have had multiple rounds of feedback and improvement, especially to the disabled states.

This was last reveiwed on 18/08 and I've incorporated all the feedback, and the base Bokeh components (which are in scope for the grant) are now ready to be implemented.

You'll find the ready-for-dev components in the [Figma file](https://www.figma.com/design/3xk3QObj3XaALcHHyz8nmE/%F0%9F%93%8A-Bokeh-Design-System?node-id=482-368&t=hPVCMnWMOUphleFT-1).

<img width="2550" height="1345" alt="Image" src="https://github.com/user-attachments/assets/a5e30815-f0aa-4320-be9f-6f1b0e52fa48" />
